File: ClojureCLR/Clojure/Clojure/Lib/AFunction.cs
/**
 *   Copyright (c) David Miller. All rights reserved.
 *   The use and distribution terms for this software are covered by the
 *   Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
 *   which can be found in the file epl-v10.html at the root of this distribution.
 *   By using this software in any fashion, you are agreeing to be bound by
 * 	 the terms of this license.
 *   You must not remove this notice, or any other, from this software.
 **/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;

namespace clojure.lang
{
    /// <summary>
    /// Base class providing IComparer implementation on top of <see cref="AFn">AFn</see>.  Internal use by compiler.
    /// </summary>
    public abstract class AFunction : AFn, Fn, IComparer
    {
        #region Ctors and factory methods

        /// <summary>
        /// Initialize an <see cref="AFunction">AFunction</see> to have the given metadata.
        /// </summary>
        /// <param name="meta">The metadata to attach.</param>
        public AFunction(IPersistentMap meta)
            : base(meta)
        {
        }

        /// <summary>
        /// Initialize an <see cref="AFunction">AFunction</see> to have the null metadata.
        /// </summary>
        public AFunction()
            : base()
        {
        }

        #endregion

        #region IComparer Members

        /// <summary>
        /// Compares two objects and returns a value indicating whether one is less than, equal to, or greater than the other.
        /// </summary>
        /// <param name="x">The first object to compare.</param>
        /// <param name="y">The second object to compare.</param>
        /// <returns>x?y:-1 if less than, 0 if equal, 1 if greater</returns>
        /// <remarks>Uses the two-parameter invoke.  
        /// Return type can be <code>bool</code> with <value>true</value> meaning less-than, or an int.</remarks>
        public int Compare(object x, object y)
        {
            Object o = invoke(x, y);

            if (o is Boolean)
            {
                if (RT.booleanCast(o))
                    return -1;
                return RT.booleanCast(invoke(y, x)) ? 1 : 0;
            }
            return Util.ConvertToInt(o);
        }

        #endregion
    
    }
}

File: src/jvm/clojure/lang/AFunction.java
/**
 *   Copyright (c) Rich Hickey. All rights reserved.
 *   The use and distribution terms for this software are covered by the
 *   Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
 *   which can be found in the file epl-v10.html at the root of this distribution.
 *   By using this software in any fashion, you are agreeing to be bound by
 * 	 the terms of this license.
 *   You must not remove this notice, or any other, from this software.
 **/

/* rich Dec 16, 2008 */

package clojure.lang;

import java.io.Serializable;
import java.util.Comparator;

public abstract class AFunction extends AFn implements IObj, Comparator, Fn, Serializable {

public volatile MethodImplCache __methodImplCache;

public IPersistentMap meta(){
	return null;
}

public IObj withMeta(final IPersistentMap meta){
	return new RestFn(){
		protected Object doInvoke(Object args) {
			return AFunction.this.applyTo((ISeq) args);
		}

		public IPersistentMap meta(){
			return meta;
		}

		public IObj withMeta(IPersistentMap meta){
			return AFunction.this.withMeta(meta);
		}

		public int getRequiredArity(){
			return 0;
		}
	};
}

public int compare(Object o1, Object o2){
	try
		{
		Object o = invoke(o1, o2);

		if(o instanceof Boolean)
			{
			if(RT.booleanCast(o))
				return -1;
			return RT.booleanCast(invoke(o2,o1))? 1 : 0;
			}

		Number n = (Number) o;
		return n.intValue();
		}
	catch(Exception e)
		{
		throw Util.sneakyThrow(e);
		}
}
}
